app/src/main/java/io/gnosis/safe/ui/safe/SafeSelectionDialog.kt

```diff
@@ -0,0 +1,277 @@
+import io.gnosis.safe.di.components.DaggerViewComponent
+import io.gnosis.safe.di.modules.ViewModule
+import io.gnosis.safe.ui.base.AppDispatchers
+import io.gnosis.safe.ui.base.BaseBottomSheetDialogFragment
+import io.gnosis.safe.ui.base.BaseStateViewModel
+import io.gnosis.safe.ui.base.UnsupportedViewType
+import io.gnosis.safe.utils.asMiddleEllipsized
+import pm.gnosis.svalinn.common.utils.visible
+import pm.gnosis.utils.asEthereumAddressString
+import java.lang.ref.WeakReference
+import javax.inject.Inject
+
+class SafeSelectionDialog : BaseBottomSheetDialogFragment<DialogSafeSelectionBinding>() {
+
+    @Inject
+    lateinit var viewModel: SafeSelectionViewModel
+
+    @Inject
+    lateinit var adapter: SafeSelectionAdapter
+
+    override fun inject() {
+        DaggerViewComponent.builder()
+            .viewModule(ViewModule(context!!))
+            .applicationComponent(HeimdallApplication[context!!])
+            .build()
+            .inject(this)
+    }
+
+
+    override fun inflateBinding(inflater: LayoutInflater, container: ViewGroup?): DialogSafeSelectionBinding =
+        DialogSafeSelectionBinding.inflate(inflater, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        with(binding) {
+            list.layoutManager = LinearLayoutManager(context)
+            list.addItemDecoration(DividerItemDecoration(context, LinearLayoutManager.VERTICAL))
+            list.adapter = adapter
+        }
+
+        viewModel.state.observe(viewLifecycleOwner, Observer { state ->
+
+            when(state) {
+                is SafeSelectionState.SafeListState -> {
+                    adapter.setItems(state.listItems, state.activeSafe)
+                }
+                is SafeSelectionState.AddSafeState -> {
+
+                    state.viewAction?.let {
+                        val action = it as BaseStateViewModel.ViewAction.NavigateTo
+                        Navigation.findNavController(activity as FragmentActivity, R.id.safe_overview_root).navigate(action.navDirections)
+                        dismiss()
+                    }
+                }
+            }
+        })
+
+        viewModel.loadSafes()
+    }
+
+    companion object {
+
+        private val TAG = SafeSelectionDialog::class.java.simpleName
+
+        fun show(context: Context) {
+            val dialog = SafeSelectionDialog()
+            dialog.show((context as FragmentActivity).supportFragmentManager, TAG)
+        }
+    }
+}
+
+sealed class SafeSelectionState : BaseStateViewModel.State {
+
+    data class SafeListState(
+        val listItems: List<Any>,
+        val activeSafe: Safe?,
+        override var viewAction: BaseStateViewModel.ViewAction?
+    ) : SafeSelectionState()
+
+    data class AddSafeState(
+        override var viewAction: BaseStateViewModel.ViewAction?
+    ) : SafeSelectionState()
+}
+
+
+class SafeSelectionViewModel @Inject constructor(
```

Comment:
I think we can put this in a separate source file like we usually do with viewModels. The viewModel and the states emitted by it.

Reply:
I think short view models that are used only in a fragment can coexist in a same file with fragment. Then everything related is in one file and you don't have to jump around. But it doesn't matter to me. We have to agree on one approach and stick to it

Reply:
Well this file is nearly 300 lines long if you don't want to split up the viewModel then something else.

Reply:
What I suggest is to do the following, and using the package structure as guidance:

ui.safe.
            add.
                  //...
            select.
                 Dialog
                 Adapter
                 ViewModel
           overview
                 fragment
                 viewModel
           Balances. //this doesn't need to be on your PR
                .nfts
                .fts